File: user-service/src/main/java/com/phorus/userservice/controllers/UserController.java
package com.phorus.userservice.controllers;

import com.phorus.userservice.mappers.UserMapper;
import com.phorus.userservice.model.dtos.UserDTO;
import com.phorus.userservice.model.dtos.responses.UserResponse;
import com.phorus.userservice.model.validationGroups.UserValidationGroup;
import com.phorus.userservice.services.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

import java.net.URI;


@RestController
@RequestMapping("/users")
public class UserController {

    private final UserService userService;
    private final UserMapper userMapper;

    @Autowired
    public UserController(UserService userService, UserMapper userMapper) {
        this.userService = userService;
        this.userMapper = userMapper;
    }

    @GetMapping("/{userId}")
    public Mono<ResponseEntity<UserResponse>> get(@PathVariable Long userId) {

        return userService.findById(userId)
                .map(userMapper::userToUserResponse)
                .map(ResponseEntity::ok);
    }

    @PostMapping
    public Mono<ResponseEntity<Long>> create(@Validated(UserValidationGroup.UserCreateValGroup.class)
                                                 @RequestBody Mono<UserDTO> userDTO) {

        return userDTO.map(userMapper::userDtoToUser)
                .flatMap(user -> userService.create(Mono.just(user)))
                .map(userId -> ResponseEntity.created(URI.create("/users/" + userId)).build());
    }

    @PutMapping
    public Mono<ResponseEntity<Void>> update(@RequestParam("userId") Long userId,
                                             @Validated(UserValidationGroup.UserUpdateValGroup.class)
                                             @RequestBody Mono<UserDTO> userDTO) {

        return userService.update(userId, userDTO)
                .flatMap(p -> Mono.just(ResponseEntity.ok().build()));
    }

    @DeleteMapping("/{userId}")
    public Mono<ResponseEntity<Void>> delete(@PathVariable("userId") Long userId) {

        return userService.deleteById(userId)
                .flatMap(p -> Mono.just(ResponseEntity.ok().build()));
    }


    @GetMapping("/getAmountOfPets")
    public Mono<ResponseEntity<Long>> getAmountOfPets(@RequestParam("userId") Long userId) {

        return userService.getAmountOfPets(userId)
                .flatMap(p -> Mono.just(ResponseEntity.ok(p)));
    }
}

File: user-service/src/main/kotlin/com/phorus/userservice/controllers/UserController.kt
package com.phorus.userservice.controllers;

import com.phorus.userservice.mappings.toEntity
import com.phorus.userservice.mappings.toResponse
import com.phorus.userservice.model.dtos.UserDTO
import com.phorus.userservice.model.dtos.UserResponse
import com.phorus.userservice.model.validationGroups.UserCreateValGroup
import com.phorus.userservice.model.validationGroups.UserUpdateValGroup
import com.phorus.userservice.services.UserService
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import java.net.URI


@RestController
@RequestMapping("/users")
class UserController(
    private val userService: UserService,
) {

    @GetMapping("/{userId}")
    suspend fun get(@PathVariable userId: Long): ResponseEntity<UserResponse> =
        userService.findById(userId)
            .toResponse()
            .let { ResponseEntity.ok(it) }

    @PostMapping
    suspend fun create(
        @Validated(UserCreateValGroup::class)
        @RequestBody
        userDTO: UserDTO
    ): ResponseEntity<Long> =
        userDTO.toEntity()
            .let { userService.create(it) }
            .let { ResponseEntity.created(URI.create("/users/$it")).build() }

    @PutMapping
    suspend fun update(
        @RequestParam("userId")
        userId: Long,

        @Validated(UserUpdateValGroup::class)
        @RequestBody
        userDTO: UserDTO
    ): ResponseEntity<Void> =
        userService.update(userId, userDTO)
            .let { ResponseEntity.ok().build() }

    @DeleteMapping("/{userId}")
    suspend fun delete(@PathVariable("userId") userId: Long): ResponseEntity<Void> =
        userService.deleteById(userId)
            .let { ResponseEntity.ok().build() }


    @GetMapping("/getAmountOfPets")
    suspend fun getAmountOfPets(@RequestParam("userId") userId: Long): ResponseEntity<Long> =
        userService.getAmountOfPets(userId)
            .let { ResponseEntity.ok(it) }
}
